primitives/src/locktime/relative.rs

```diff
@@ -251,60 +233,110 @@ impl LockTime {
+                Ok(required_height <= last_block_height.to_consensus_u32())
+            }
+            Time(time) => Err(IncompatibleHeightError(time.value())),
         }
     }
 
     /// Returns true if this [`relative::LockTime`] is satisfied by [`Time`].
     ///
+    /// > The relative lock-time specifies a timespan in units of 512 seconds granularity. The
+    /// > timespan starts from the median-time-past of the output’s previous block, and ends at the
+    /// > MTP of the previous block.
+    /// >
+    /// > ... a relative time-based lock-time n can be included into any block produced 512 * n
+    /// > seconds after the mining date of the output it is spending, or any block thereafter. The
+    /// > mining date of the output is equal to the median-time-past of the previous block which
+    /// > mined it.
+    ///
+    /// > The block produced time is equal to the median-time-past of its previous block.
+    ///
+    /// # Parameters
+    ///
+    /// * `prevout_mining_date`: The mining date of the output (defined in quote above).
+    /// * `block_produced_time`: The MTP of the previous block (defined in quote above).
+    ///
+    /// Both parameters are UNIX timestamps.
+    ///
     /// # Errors
     ///
     /// Returns an error if this lock is not lock-by-time.
     ///
     /// # Examples
     ///
     /// ```rust
-    /// # use bitcoin_primitives::Sequence;
-    /// # use bitcoin_primitives::locktime::relative::{LockTime, Height, Time};
+    /// use bitcoin_primitives::Sequence;
+    /// use units::locktime::absolute::Time;
+    ///
+    /// // Units of 512 seconds that the output we want to spend has to wait.
+    /// let span: u16 = 100;
     ///
-    /// let intervals: u16 = 70; // approx 10 hours;
-    /// let lock = Sequence::from_512_second_intervals(intervals).to_relative_lock_time().expect("valid time");
-    /// assert!(lock.is_satisfied_by_time(Time::from_512_second_intervals(intervals + 10)).expect("a time"));
+    /// // Arbitrary UNIX timestamp representing the MTP of the block that created the prevout we want to spend.
+    /// let prevout_mining_date = Time::from_consensus(1_000_000_000).expect("valid time");
+    ///
+    /// // A unix timestamp 512 seconds after the timespan since the MTP of the prevout block.
+    /// let mtp_of_tip_of_chain = Time::from_consensus(1_000_000_000 + 512 * span as u32).expect("valid time");
+    ///
+    /// let lock = Sequence::from_512_second_intervals(span).to_relative_lock_time().expect("valid interval");
+    /// assert!(lock.is_satisfied_by_time(prevout_mining_date, mtp_of_tip_of_chain).unwrap());
     /// ```
     #[inline]
     #[cfg_attr(all(test, mutate), mutate)]
-    pub fn is_satisfied_by_time(&self, time: Time) -> Result<bool, IncompatibleTimeError> {
+    pub fn is_satisfied_by_time(
+        &self,
+        prevout_mining_date: Time,
+        chain_mtp: Time,
+    ) -> Result<bool, IncompatibleTimeError> {
```

Comment:
Changed the types of the inputs to the absolute ones... Since they match what we want.

At least the consumer will not mess the height with a unix timestamp.

Reply:
>The best solution may be to just live with the two u32s. If the user gets them mixed up the function will be wrong 100% of the time which hopefully will lead to them quickly noticing and fixing it.

What you think about this ? @apoelstra

Reply:
How does one get an `abosulet::Time` out of a blockheader?

Reply:
Really what I want is the ability to subtract two blocktimes to get a `relative::LockTime`, but

* `relative::LockTime` does not have enough range (the difference between two blocktimes can span the full range of `u32`, which is much more than can be specified as a relative locktime)
* this still leads us to an easy type confusion, this time between the entailing/entailee locktimes, rather than the early/later blocktimes.

Also the times in the blockheaders **are** UNIX timestamps represented as `u32`s. That's just what's on the chain. I think it causes confusion to try to abstract that away.

Reply:
> How does one get an abosulet::Time out of a blockheader?

`block.header.time` ?

>the difference between two blocktimes can span the full range of u32, which is much more than can be specified as a relative locktime

Yes, thats why we store them as 512 intervals in a u16 camp.
from the header would be something like.

```Rust

let timestamp: u32 = block.header.time

let other_timestamp: u32 = block.header.time + 1_000_000_000

let time_to_wait: u16 = ((other_timestamp - timestamp) / 512) as u16 // we need the cast

let sequence = Sequence::from_512_intervals(time_to_wait);  
```

>We could try something addition based, e.g.
>Other than that, I think the original API was the best one.

I dont think that where talking about the same thing. Im really confused about what are you talking, im sorry but can you re-word your question about the change ? Like, the old api is wrong, its not working.

Reply:
> `block.header.time` ?

This is a `u32` containing a UNIX timestamp, not a `Time`.